File: api/torch/weight_synchronizer_test.py
# Copyright 2026 Google LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""E2E physical integration tests for PyTorch WeightSynchronizer on XLA TPUs."""

import os
import time

from absl.testing import absltest
from absl.testing import parameterized
import numpy as np
import torch
import torch_tpu

from api.torch.weight_synchronizer import WeightSynchronizer


class WeightSynchronizerTorchTest(parameterized.TestCase):

  def setUp(self):
    super().setUp()
    # Initialize PyTorch XLA accelerator device E2E
    self.device = torch.device("tpu")
    self.num_layers = 2
    self.num_shards = 1
    self.block_size = 2
    self.slice_byte_size = 16384 // 4  # float32 capacity

  @parameterized.named_parameters(
      ("fp32", torch.float32),
      ("int32", torch.int32),
  )
  def test_e2e_3node_distributed_weight_push_and_pull(self, dtype):
    shape = (self.block_size, 128, 8)  # 16384 bytes capacity per layer shard

    # 1. Allocate source (Trainer) weights on Device TPU
    src_tensors = []
    for l in range(self.num_layers):
      shards = []
      for sh in range(self.num_shards):
        t = torch.zeros(shape, dtype=dtype, device=self.device)
        shards.append(t)
      src_tensors.append(shards)

    # Allocate destination 1 (Inference Peer 1) weights
    dst1_tensors = []
    for l in range(self.num_layers):
      shards = []
      for sh in range(self.num_shards):
        t = torch.zeros(shape, dtype=dtype, device=self.device)
        shards.append(t)
      dst1_tensors.append(shards)

    # Allocate destination 2 (Inference Peer 2) weights
    dst2_tensors = []
    for l in range(self.num_layers):
      shards = []
      for sh in range(self.num_shards):
        t = torch.zeros(shape, dtype=dtype, device=self.device)
        shards.append(t)
      dst2_tensors.append(shards)

    # 2. Instantiate destination WeightSynchronizers on ephemeral ports!
    ws_dest1 = WeightSynchronizer(dst1_tensors, local_port=0, parallelism=1)
    ws_dest2 = WeightSynchronizer(dst2_tensors, local_port=0, parallelism=1)

    self.assertIsNotNone(ws_dest1.local_port)
    self.assertIsNotNone(ws_dest2.local_port)

    peer_dest1 = f"localhost:{ws_dest1.local_port}"
    peer_dest2 = f"localhost:{ws_dest2.local_port}"

    # ==========================================================================
    # Scenario A: Test the Push API E2E (1 Source pushes to 2 Destinations!)
    # ==========================================================================
    # Trainer fills source weights with distinct values per layer
    for l in range(self.num_layers):
      for sh in range(self.num_shards):
        val = float(l + 10.0)  # Layer 0=10.0, Layer 1=11.0
        src_tensors[l][sh].fill_(val)

    # Recreate/Instantiate ws_source to capture filled buffers!
    ws_source = WeightSynchronizer(src_tensors, local_port=0, parallelism=1)
    self.assertIsNotNone(ws_source.local_port)
    peer_source = f"localhost:{ws_source.local_port}"

    # Source pushes weights to both dest1 and dest2 socket servers E2E!
    ws_source.push_weights([peer_dest1, peer_dest2])

    # Assert both destinations have received the trainer's weights on their TPU HBM!
    for l in range(self.num_layers):
      for sh in range(self.num_shards):
        expected_val = float(l + 10.0)
        np.testing.assert_allclose(
            dst1_tensors[l][sh].cpu().numpy(), expected_val, atol=1e-5
        )
        np.testing.assert_allclose(
            dst2_tensors[l][sh].cpu().numpy(), expected_val, atol=1e-5
        )

    # ==========================================================================
    # Scenario B: Test the Pull API E2E (2 Destinations pull from 1 Source!)
    # ==========================================================================
    # Trainer updates source weights with brand new values
    for l in range(self.num_layers):
      for sh in range(self.num_shards):
        val = float(l + 20.0)  # Layer 0=20.0, Layer 1=21.0
        src_tensors[l][sh].fill_(val)

    # Recreate ws_source to capture new buffers!
    ws_source = WeightSynchronizer(src_tensors, local_port=0, parallelism=1)
    self.assertIsNotNone(ws_source.local_port)
    peer_source = f"localhost:{ws_source.local_port}"

    # Self-push to populate host buffer!
    ws_source.push_weights([peer_source])

    # Dest1 and Dest2 independently pull weights from the trainer source!
    ws_dest1.pull_weights(peer_source)
    ws_dest2.pull_weights(peer_source)

    # Assert both destinations have successfully pulled the trainer's new weights E2E!
    for l in range(self.num_layers):
      for sh in range(self.num_shards):
        expected_val = float(l + 20.0)
        np.testing.assert_allclose(
            dst1_tensors[l][sh].cpu().numpy(), expected_val, atol=1e-5
        )
        np.testing.assert_allclose(
            dst2_tensors[l][sh].cpu().numpy(), expected_val, atol=1e-5
        )


if __name__ == "__main__":
  absltest.main()

File: weight_sync/weight_synchronizer_test.cc
// Copyright 2026 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <cstddef>
#include <cstdint>
#include <cstring>
#include <memory>
#include <string>
#include <vector>

#include <gtest/gtest.h>
#include "absl/log/log.h"
#include "absl/status/status.h"
#include "weight_sync/weight_synchronizer_base.h"

namespace tpu_raiden {
namespace weight_sync {
namespace {

class WeightSynchronizerTest : public ::testing::Test {
 protected:
  void SetUp() override {
    // Unified physical test parameters representing 64KB weight buffer E2E!
    num_layers_ = 1;
    num_shards_ = 1;
    slice_byte_size_ = 65536;  // 64KB
  }

  size_t num_layers_;
  size_t num_shards_;
  size_t slice_byte_size_;
};

TEST_F(WeightSynchronizerTest, PushAndPullWeightsCorrectnessE2e) {
  // 1. Instantiate three independent CPU-only synchronizers locally E2E!
  // ws_source represents the active RL Trainer
  auto ws_source = std::make_unique<WeightSynchronizerBase>(
      num_layers_, num_shards_, slice_byte_size_,
      /*local_port=*/0, /*host_blocks_to_allocate=*/1);

  // ws_dest1 and ws_dest2 represent inference server peers
  auto ws_dest1 = std::make_unique<WeightSynchronizerBase>(
      num_layers_, num_shards_, slice_byte_size_,
      /*local_port=*/0, /*host_blocks_to_allocate=*/1);

  auto ws_dest2 = std::make_unique<WeightSynchronizerBase>(
      num_layers_, num_shards_, slice_byte_size_,
      /*local_port=*/0, /*host_blocks_to_allocate=*/1);

  ASSERT_TRUE(ws_source->local_port().has_value());
  ASSERT_TRUE(ws_dest1->local_port().has_value());
  ASSERT_TRUE(ws_dest2->local_port().has_value());

  std::string source_peer =
      "localhost:" + std::to_string(*ws_source->local_port());
  std::string dest1_peer =
      "localhost:" + std::to_string(*ws_dest1->local_port());
  std::string dest2_peer =
      "localhost:" + std::to_string(*ws_dest2->local_port());

  LOG(INFO) << "Launched C++ Weight Syncers: Source=" << source_peer
            << ", Dest1=" << dest1_peer << ", Dest2=" << dest2_peer;

  // 2. Populate the Trainer source buffer with distinct byte pattern (0xAB)
  // E2E!
  uint8_t* src_host_ptr = const_cast<uint8_t*>(ws_source->GetHostPointer(0, 0));
  ASSERT_NE(src_host_ptr, nullptr);
  std::memset(src_host_ptr, 0xAB, slice_byte_size_);

  // Populate inference servers with zeros baseline
  uint8_t* dest1_host_ptr =
      const_cast<uint8_t*>(ws_dest1->GetHostPointer(0, 0));
  uint8_t* dest2_host_ptr =
      const_cast<uint8_t*>(ws_dest2->GetHostPointer(0, 0));
  ASSERT_NE(dest1_host_ptr, nullptr);
  ASSERT_NE(dest2_host_ptr, nullptr);
  std::memset(dest1_host_ptr, 0x00, slice_byte_size_);
  std::memset(dest2_host_ptr, 0x00, slice_byte_size_);

  // Assert baseline state (zeros)
  EXPECT_EQ(dest1_host_ptr[0], 0x00);
  EXPECT_EQ(dest2_host_ptr[0], 0x00);

  // ==========================================================================
  // Test Scenario 1: Push weights from Trainer ws_source to both inference
  // peers!
  // ==========================================================================
  absl::Status push_status = ws_source->PushWeights({dest1_peer, dest2_peer});
  ASSERT_TRUE(push_status.ok()) << push_status.message();

  // Assert successful E2E network sockets streaming and exact byte parity!
  for (size_t i = 0; i < slice_byte_size_; ++i) {
    EXPECT_EQ(dest1_host_ptr[i], 0xAB) << "Mismatch at byte " << i;
    EXPECT_EQ(dest2_host_ptr[i], 0xAB) << "Mismatch at byte " << i;
  }

  // ==========================================================================
  // Test Scenario 2: Pull weights from Trainer ws_source E2E!
  // ==========================================================================
  // Reset destinations back to zeros baseline
  std::memset(dest1_host_ptr, 0x00, slice_byte_size_);
  std::memset(dest2_host_ptr, 0x00, slice_byte_size_);

  // ws_dest1 and ws_dest2 pull current weights from the trainer source peer!
  absl::Status pull_status1 = ws_dest1->PullWeights(source_peer);
  absl::Status pull_status2 = ws_dest2->PullWeights(source_peer);
  ASSERT_TRUE(pull_status1.ok()) << pull_status1.message();
  ASSERT_TRUE(pull_status2.ok()) << pull_status2.message();

  // Assert successful pulls E2E and exact byte parity!
  for (size_t i = 0; i < slice_byte_size_; ++i) {
    EXPECT_EQ(dest1_host_ptr[i], 0xAB) << "Mismatch at byte " << i;
    EXPECT_EQ(dest2_host_ptr[i], 0xAB) << "Mismatch at byte " << i;
  }
}

}  // namespace
}  // namespace weight_sync
}  // namespace tpu_raiden
